core/account/utxodb/reserve.go

```diff
@@ -0,0 +1,365 @@
+package utxodb
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"chain/database/pg"
+	"chain/errors"
+	"chain/protocol/bc"
+	"chain/sync/idempotency"
+)
+
+var (
+	// ErrInsufficient indicates the account doesn't contain enough
+	// units of the requested asset to satisfy the reservation.
+	// New units must be deposited into the account in order to
+	// satisfy the request; change will not be sufficient.
+	ErrInsufficient = errors.New("reservation found insufficient funds")
+
+	// ErrReserved indicates that a reservation could not be
+	// satisfied because some of the outputs were already reserved.
+	// When those reservations are finalized into a transaction
+	// (and no other transaction spends funds from the account),
+	// new change outputs will be created
+	// in sufficient amounts to satisfy the request.
+	ErrReserved = errors.New("reservation found outputs already reserved")
+)
+
+// UTXO describes an individual account UTXO.
+type UTXO struct {
+	bc.Outpoint
+	bc.AssetAmount
+	Script []byte
+
+	AccountID           string
+	ControlProgramIndex uint64
+}
+
+// Change represents reserved units beyond what was asked for.
+// Total reservation is for Amount+Source.Amount.
+type Change struct {
+	Source Source
+	Amount uint64
+}
+
+// Source describes the criteria to use when selecting UTXOs.
+type Source struct {
+	AssetID     bc.AssetID `json:"asset_id"`
+	AccountID   string     `json:"account_id"`
+	Amount      uint64
+	ClientToken *string `json:"client_token"`
+}
+
+// Reservation describes a reservation of a set of UTXOs belonging
+// to a particular account. Reservations are immutable.
+type Reservation struct {
+	ID          uint64
+	AccountID   string
+	UTXOs       []*UTXO
+	Change      []Change
```

Comment:
See https://github.com/chain/chain/pull/138. You can simplify `Change` to be a single `uint64` (among other improvements).